src/appleseed/foundation/math/microfacet2.h

```diff
@@ -472,29 +507,153 @@ class BeckmannMDF2
+                *this,
+                v,
+                s,
+                alpha_x,
+                alpha_y);
+    }
+
+    virtual T do_eval_pdf(
+        const Vector<T, 3>&  v,
         const Vector<T, 3>&  h,
         const T              alpha_x,
         const T              alpha_y) const APPLESEED_OVERRIDE
     {
         return
-            BeckmannSmithMaskingShadowing<T>::G(
-                incoming,
-                outgoing,
+            MDF<T>::do_eval_Dw(
+                *this,
+                v,
                 h,
                 alpha_x,
                 alpha_y);
     }
+
+    // This code comes from OpenShadingLanguage test render.
+    Vector<T, 2> do_sample_11(
+        const T             cos_theta,
+        const Vector<T, 3>& s) const
+    {
+        const T ct = std::max(cos_theta, T(1e-6));
+        const T tan_theta = std::sqrt(T(1.0) - square(ct)) / ct;
+        const T cot_theta = T(1.0) / tan_theta;
+
+        // Sample slope X:
+        // compute a coarse approximation using the approximation:
+        // exp(-ierf(x)^2) ~= 1 - x * x
+        // solve y = 1 + b + K * (1 - b * b).
+
+        const T c = erf(cot_theta);
+        const T K = tan_theta / std::sqrt(Pi);
+        const T y_approx = s[0] * (T(1.0) + c + K * (1 - c * c));
+        const T y_exact  = s[0] * (T(1.0) + c + K * std::exp(-square(cot_theta)));
+        T b = K > T(0.0)
+            ? (T(0.5) - std::sqrt(K * (K - y_approx + T(1.0)) + T(0.25))) / K
+            : y_approx - T(1.0);
+
+        // Perform newton step to refine toward the true root.
+        T inv_erf = erf_inv(b);
+        T value  = T(1.0) + b + K * std::exp(-square(inv_erf)) - y_exact;
+
+        // Check if we are close enough already.
+        // This also avoids NaNs as we get close to the root.
+        Vector2d slope;
+
+        if (std::fabs(value) > T(1e-6))
+        {
+            b -= value / (T(1.0) - inv_erf * tan_theta); // newton step 1
+            T inv_erf = erf_inv(b);
+            value = T(1.0) + b + K * std::exp(-square(inv_erf)) - y_exact;
+            b -= value / (T(1.0) - inv_erf * tan_theta); // newton step 2
+            // Compute the slope from the refined value
+            slope[0] = erf_inv(b);
+        }
+        else
+        {
+            // we are close enough already
+            slope[0] = inv_erf;
+        }
+
+        // Sample slope Y
+        slope[1] = erf_inv(T(2.0) * s[1] - T(1.0));
+        return slope;
+    }
+
+    typedef boost::math::policies::policy<
+#ifndef NDEBUG
+//       boost::math::policies::error_policy_type<boost::math::policies::ignore_error>,
+#endif
+        boost::math::policies::promote_float<false>,
+        boost::math::policies::digits10<6>
+    > ErfPolicyType;
+
+    static T erf(const T x)
+    {
+        return boost::math::erf(x, ErfPolicyType());
+    }
+
+    static T erf_inv(T x)
+    {
+        if (std::fabs(x) == T(1.0))
+            x -= boost::math::copysign(T(1e-5), x);
```

Comment:
You're trying to have a value slightly less than 1.0 in absolute value, right? Wouldn't multiplying by (1.0 - 1.0e-5) be enough? Or in this case, since you know the value is exactly 1.0, just replace by by 0.99999.